### PR TITLE
Feature/ReturnCode [6099]

### PR DESCRIPTION
**Pull request by laura-eprosima**
Change return type to ReturnCode in those functions that need it according to the DDS standard

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/136/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/125/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/135/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/140/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/129/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/139/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/248/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/226/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/247/)

File: include/fastdds/dds/subscriber/Subscriber.hpp [hunk 1]
```diff
@@ -24,6 +24,9 @@
 
 #include <fastdds/dds/topic/DataReaderListener.hpp>
 #include <fastdds/dds/subscriber/qos/SubscriberQos.hpp>
+#include <fastrtps/types/TypesBase.h>
+
+using namespace eprosima::fastrtps::types;
```
Comment (LuisGP) [inline include/fastdds/dds/subscriber/Subscriber.hpp]: I'll not comment this again :).

File: include/fastdds/rtps/builtin/discovery/participant/PDP.h [hunk 2]
```diff
@@ -80,6 +80,7 @@ class PDP
     /**
      * Initialize the PDP.
      * @param part Pointer to the RTPSParticipant.
+     * @param enableReader bool that indicates if DATA(p) reception should be activated
```
Comment (LuisGP) [inline include/fastdds/rtps/builtin/discovery/participant/PDP.h]: Does this change belong to this PR?

File: src/cpp/dds/domain/DomainParticipantImpl.cpp [hunk 3]
```diff
@@ -135,38 +135,58 @@ DomainParticipantImpl::~DomainParticipantImpl()
...
     auto pit = publishers_.find(pub);
 
     if (pit != publishers_.end() && pub->get_instance_handle() == pit->second->get_instance_handle())
     {
+        std::vector<DataWriter*> writers;
+        pub->get_datawriters(writers);
+        if(!writers.empty())
+        {
+            return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+        }
```
Comment (LuisGP) [inline src/cpp/dds/domain/DomainParticipantImpl.cpp]: Commented in PR #634

File: src/cpp/dds/domain/DomainParticipantImpl.cpp [hunk 4]
```diff
@@ -135,38 +135,58 @@ DomainParticipantImpl::~DomainParticipantImpl()
...
     auto sit = subscribers_.find(sub);
 
     if (sit != subscribers_.end() && sub->get_instance_handle() == sit->second->get_instance_handle())
     {
+        std::vector<DataReader*> readers;
+        sub->get_datareaders(readers);
+        if(!readers.empty())
+        {
+            return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+        }
```
Comment (LuisGP) [inline src/cpp/dds/domain/DomainParticipantImpl.cpp]: PR #634

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/258/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/235/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/257/)

File: src/cpp/dds/domain/DomainParticipantImpl.cpp (hunk 4 above)
Comment (laura-eprosima) [reply src/cpp/dds/domain/DomainParticipantImpl.cpp]: Done

File: src/cpp/dds/domain/DomainParticipantImpl.cpp (hunk 3 above)
Comment (laura-eprosima) [reply src/cpp/dds/domain/DomainParticipantImpl.cpp]: Done

File: include/fastdds/dds/subscriber/Subscriber.hpp (hunk 1 above)
Comment (laura-eprosima) [reply include/fastdds/dds/subscriber/Subscriber.hpp]: Done

File: include/fastdds/dds/publisher/Publisher.hpp [hunk 5]
```diff
@@ -23,6 +23,9 @@
 #include <fastrtps/fastrtps_dll.h>
 #include <fastdds/rtps/common/Time_t.h>
 #include <fastrtps/attributes/PublisherAttributes.h>
+#include <fastrtps/types/TypesBase.h>
+
+using namespace eprosima::fastrtps::types;
```
[the remark this answers is not in the document]
Comment (laura-eprosima) [reply include/fastdds/dds/publisher/Publisher.hpp]: Done

File: include/fastdds/dds/domain/DomainParticipantFactory.hpp [hunk 6]
```diff
@@ -21,9 +21,13 @@
 #define _FASTDDS_DOMAINPARTICIPANT_HPP_
 
 #include <fastrtps/attributes/ParticipantAttributes.h>
+#include <fastrtps/types/TypesBase.h>
+
 #include <mutex>
 #include <map>
 
+using namespace eprosima::fastrtps::types;
```
[the remark this answers is not in the document]
Comment (laura-eprosima) [reply include/fastdds/dds/domain/DomainParticipantFactory.hpp]: Done

File: include/fastdds/dds/domain/DomainParticipant.hpp [hunk 7]
```diff
@@ -24,9 +24,12 @@
 
 #include <fastdds/rtps/common/Guid.h>
 #include <fastdds/rtps/attributes/RTPSParticipantAttributes.h>
+#include <fastrtps/types/TypesBase.h>
 
 #include <utility>
 
+using namespace eprosima::fastrtps::types;
```
[the remark this answers is not in the document]
Comment (laura-eprosima) [reply include/fastdds/dds/domain/DomainParticipant.hpp]: Done

File: include/fastdds/rtps/builtin/discovery/participant/PDP.h (hunk 2 above)
Comment (laura-eprosima) [reply include/fastdds/rtps/builtin/discovery/participant/PDP.h]: No and I don't know where it comes

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/331/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/304/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/329/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/371/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/338/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/369/)

File: src/cpp/dds/topic/DataReaderImpl.hpp [hunk 8]
```diff
@@ -33,6 +33,9 @@
 #include <fastdds/rtps/reader/ReaderListener.h>
 #include <fastrtps/attributes/TopicAttributes.h>
 #include <fastrtps/qos/LivelinessChangedStatus.h>
+#include <fastrtps/types/TypesBase.h>
+
+using namespace eprosima::fastrtps::types;
```
Comment (laura-eprosima) [inline src/cpp/dds/topic/DataReaderImpl.hpp]:
```suggestion
using eprosima::fastrtps::types::ReturnCode_t;
```

File: src/cpp/dds/topic/DataWriterImpl.cpp [hunk 9]
```diff
@@ -154,22 +154,38 @@ bool DataWriterImpl::write(
...
+    }
+    return ReturnCode_t::RETCODE_ERROR;
 }
 
-bool DataWriterImpl::dispose(
+ReturnCode_t DataWriterImpl::dispose(
         void* data,
         const rtps::InstanceHandle_t& handle)
 {
+    if(!handle.isDefined())
```
Comment (laura-eprosima) [inline src/cpp/dds/topic/DataWriterImpl.cpp]:
```suggestion
    if (!handle.isDefined())
```

File: src/cpp/dds/topic/DataWriterImpl.hpp [hunk 10]
```diff
@@ -35,6 +35,9 @@
 #include <fastrtps/qos/DeadlineMissedStatus.h>
 #include <fastrtps/qos/IncompatibleQosStatus.hpp>
 #include <fastrtps/qos/LivelinessLostStatus.h>
+#include <fastrtps/types/TypesBase.h>
+
+using namespace eprosima::fastrtps::types;
```
Comment (laura-eprosima) [inline src/cpp/dds/topic/DataWriterImpl.hpp]:
```suggestion
using eprosima::fastrtps::types::ReturnCode_t;
```

File: src/cpp/dds/subscriber/SubscriberImpl.hpp [hunk 11]
```diff
@@ -25,16 +25,20 @@
 
 #include <fastdds/dds/topic/DataReaderListener.hpp>
 #include <fastdds/dds/subscriber/qos/SubscriberQos.hpp>
+#include <fastrtps/types/TypesBase.h>
 
 #include <mutex>
 #include <map>
 
+using namespace eprosima::fastrtps::types;
```
Comment (laura-eprosima) [inline src/cpp/dds/subscriber/SubscriberImpl.hpp]:
```suggestion
using eprosima::fastrtps::types::ReturnCode_t;
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/532/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/473/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/528/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/570/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/497/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/566/)

File: src/cpp/dds/subscriber/SubscriberImpl.cpp [hunk 12]
```diff
@@ -89,31 +89,35 @@ const SubscriberQos& SubscriberImpl::get_qos() const
...
 }
 
-bool SubscriberImpl::set_listener(
+ReturnCode_t SubscriberImpl::set_listener(
         SubscriberListener* listener)
 {
     if (listener_ == listener)
     {
-        return false;
+        return ReturnCode_t::RETCODE_ERROR;
```
Comment (LuisGP) [inline src/cpp/dds/subscriber/SubscriberImpl.cpp]: Same question as with `PublisherImpl::set_listener`.

File: src/cpp/dds/topic/DataWriterImpl.cpp [hunk 13]
```diff
@@ -154,22 +154,38 @@ bool DataWriterImpl::write(
...
 
-bool DataWriterImpl::write(
+ReturnCode_t DataWriterImpl::write(
             void* data,
             const rtps::InstanceHandle_t& handle)
 {
+    if (!handle.isDefined())
+    {
+        return ReturnCode_t::RETCODE_BAD_PARAMETER;
+    }
```
Comment (LuisGP) [inline src/cpp/dds/topic/DataWriterImpl.cpp]: The standard talks about a `HANDLE_NIL` that we don't support yet. We define a `c_InstanceHandle_Unknown` for other purposes. If we add support for `HANDLE_NIL` this check may be removed.

Comment (laura-eprosima) [reply src/cpp/dds/topic/DataWriterImpl.cpp]: TODO comment added

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/593/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/516/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/589/)

**Comment by raquelalvarezbanos** [comment]
All builds passed, see [here](http://ci.intranet.eprosima.com:8080/job/FastRTPS%20Manual/612/)

**Pull request merged**